src/windows.rs

```diff
@@ -0,0 +1,343 @@
+
+/// List of forbidden signals.
+///
+/// Some signals are impossible to replace according to POSIX and some are so special that this
+/// library refuses to handle them (eg. SIGSEGV). The routines panic in case registering one of
+/// these signals is attempted.
+///
+/// See [`register`](fn.register.html).
+pub const FORBIDDEN: &[c_int] = &[SIGKILL, SIGSTOP, SIGILL, SIGFPE, SIGSEGV];
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+struct ActionId(u64);
+
+/// An ID of registered action.
+///
+/// This is returned by all the registration routines and can be used to remove the action later on
+/// with a call to [`unregister`](fn.unregister.html).
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct SigId {
+    signal: c_int,
+    action: ActionId,
+}
+
+type Action = Fn() + Send + Sync;
+
+struct GlobalData {
+    // RCU isn't necessary here. Unlike posix signals, Windows Console handler is called in a separate thread.
+    state: Mutex<GlobalDataState>,
+}
+
+struct GlobalDataState {
+    action_id: u64,
+    all_signals: HashMap<c_int, BTreeMap<ActionId, Box<Action>>>,
+}
+
+impl GlobalData {
+    fn ensure() -> &'static Self {
+        use std::cell::UnsafeCell;
+        use std::sync::Once;
+
+        struct GlobalCell(UnsafeCell<Option<GlobalData>>);
+        unsafe impl Sync for GlobalCell {}
+        static GLOBAL: GlobalCell = GlobalCell(UnsafeCell::new(None));
+        static GLOBAL_ONCE: Once = Once::new();
+
+        GLOBAL_ONCE.call_once(|| unsafe {
+            *GLOBAL.0.get() = Some(GlobalData::new());
+        });
+
+        if let Some(ref data) = *unsafe { &*GLOBAL.0.get() } {
+            data
+        } else {
+            unreachable!();
+        }
+    }
+
+    fn new() -> Self {
+        assert!(unsafe { SetConsoleCtrlHandler(Some(handler), TRUE) != 0 });
+        GlobalData {
+            state: Mutex::new(GlobalDataState::new()),
+        }
+    }
+}
+impl GlobalDataState {
+    fn new() -> Self {
+        Self {
+            action_id: 0,
+            all_signals: HashMap::new(),
+        }
+    }
+}
+
+extern "system" fn handler(ctrl_type: DWORD) -> BOOL {
+    if let Some(signal) = to_signal(ctrl_type) {
+        let data = GlobalData::ensure();
+        let state = data.state.lock().unwrap();
+        if let Some(all_signals) = state.all_signals.get(&signal) {
+            for action in all_signals.values() {
+                action();
+            }
+
+            // returning TRUE here has two effects:
+            // 1. Default behavior is prevented.
+            // 2. Handlers registered earlier than this are not invoked.
+            //
+            // Ideally we want to prevent default behavior while keeping
+            // old callbacks, but it's impossible.
+            //
+            // And we may want to return FALSE if `all_signals` is empty.
+            // I didn't do so to align with the unix implementation.
+            return TRUE;
```

Comment:
What is the default behaviour? Because, this is actually not aligned with the unix implementation where the default is prevented, but previous signal handlers are called.

Reply:
https://docs.microsoft.com/en-us/windows/console/setconsolectrlhandler

> When a console process receives any of the control signals, its handler functions are called on a last-registered, first-called basis until one of the handlers returns TRUE. If none of the handlers returns TRUE, the default handler is called.

Therefore, TRUE prevents both previous handlers and default handlers. FALSE calls both previous handlers and default handlers (unless one of these "previous handlers" returns TRUE)